src/generic/PolyRingHom.jl

```diff
@@ -0,0 +1,262 @@
+  end
+end
+
+# Now compose with arbitrary maps
+
+# I technically cannot do the Nothing version
+
+# I can only do the Map version of the coefficient map has codomain C
+function compose(F::PolyRingAnyMap{D, C, <: Map, <: Any}, G::S) where {D, C, S <: Map{C, <: Any}}
+  codomain(F) === domain(G) || error("Incompatible (co)domain in composition")
+  f = _coefficient_map(F)
+  if typeof(codomain(f)) === C
+    newcoeffmap = compose(f, G)
+    return hom(domain(F), codomain(G), newcoeffmap, G.(_image(F)))
+  else
+    return Generic.CompositeMap(F, G)
+  end
+end
+
+function compose(F::PolyRingAnyMap{D, C, <: Map, <: Any}, G::S) where {D, C, S <: Generic.IdentityMap{C}}
+  codomain(F) === domain(G) || error("Incompatible (co)domain in composition")
+  f = _coefficient_map(F)
+  if typeof(codomain(f)) === C
+    newcoeffmap = compose(f, G)
+    return hom(domain(F), codomain(G), newcoeffmap, G.(_image(F)))
+  else
+    return Generic.CompositeMap(F, G)
+  end
+end
+
+################################################################################
+#
+#  Types computers
+#
+################################################################################
+
+function morphism_type(::Type{D}, ::Type{C}) where {D <: PolyRing, C <: NCRing}
+  return PolyRingAnyMap{D, C, Nothing, elem_type(C)}
+end
+
+morphism_type(::D, ::C) where {D <: PolyRing, C <: NCRing} = morphism_type(D, C)
+
+function morphism_type(::Type{D}, ::Type{C}, f::Type{F}) where {D <: PolyRing, C <: NCRing, F}
+  return PolyRingAnyMap{D, C, F, elem_type(C)}
+end
+
+morphism_type(::D, ::C, ::F) where {D <: PolyRing, C <: NCRing, F} = morphism_type(D, C, F)
+
+################################################################################
+#
+#  Constructor
+#
+################################################################################
+
+@doc raw"""
+    hom(R::PolyRing, S::NCRing, [coeff_map,] image)
+    
+Given a homomorphism `coeff_map` from `C` to `S`, where `C` is the 
+coefficient ring of `R`, and given an element `image` of `S`, return the
+homomorphism from `R` to `S` whose restriction 
+to `C` is `coeff_map`, and which sends the generator of `R` to `image`.
+ 
+If no coefficient map is entered, invoke a canonical homomorphism of `C`
+to `S`, if such a homomorphism exists, and throw an error, otherwise.
+
+# Examples
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> Zx, x = ZZ["x"];
+
+julia> F = hom(Zx, Zx, x + 1);
+
+julia> F(x^2)
+x^2 + 2*x + 1
+
+julia> Fp = GF(3); Fpy, y = Fp["y"];
+
+julia> G = hom(Zx, Fpy, c -> Fp(c), y^3);
+
+julia> G(5*x + 1)
+2*y^3 + 1
+```
+"""
+function hom(R::PolyRing, S::NCRing, coeff_map, image)
+  # Now coerce into S or throw an error if not possible
+  img = _coerce(S, image)
+  return PolyRingAnyMap(R, S, coeff_map, img)
+end
+
+function hom(R::PolyRing, S::NCRing, image)
+  # Now coerce into S or throw an error if not possible
+  img = _coerce(S, image)
+  return PolyRingAnyMap(R, S, nothing, img)
```

Comment:
Hmm, shouldn't this be `Nothing`? Or it doesn't matter? If it matters, then perhaps a test could be added that would catch this?
```suggestion
  return PolyRingAnyMap(R, S, Nothing, img)
```

Reply:
I think this is correct as is

Reply:
The argument is `nothing` and not the `Nothing` type. It does matter and it is tested by the tests.